core/common/src/main/java/alluxio/collections/IndexedSet.java

```diff
@@ -131,27 +129,24 @@
    * @param otherFields other fields to index the set
    */
   public IndexedSet(FieldIndex<T> field, FieldIndex<T>... otherFields) {
-    mIndexMap = new HashMap<FieldIndex<T>, Integer>(otherFields.length + 1);
-    mIndexMap.put(field, 0);
-    for (int i = 1; i <= otherFields.length; i++) {
-      mIndexMap.put(otherFields[i - 1], i);
-    }
-
-    mSetIndexedByFieldValue = new ArrayList<Map<Object, Set<T>>>(mIndexMap.size());
-    for (int i = 0; i < mIndexMap.size(); i++) {
-      mSetIndexedByFieldValue.add(new HashMap<Object, Set<T>>());
+    Map<FieldIndex<T>, ConcurrentHashMap<Object, ConcurrentHashSet<T>>> indexMap =
+        new HashMap<>(otherFields.length + 1);
+    indexMap.put(field, new ConcurrentHashMap<Object, ConcurrentHashSet<T>>());
+    for (FieldIndex<T> fieldIndex : otherFields) {
+      indexMap.put(fieldIndex, new ConcurrentHashMap<Object, ConcurrentHashSet<T>>());
     }
+    // read only, so it is thread safe and allows concurrent access.
+    mIndexMap = Collections.unmodifiableMap(indexMap);
   }
 
   /**
    * Removes all the entries in this set.
+   *
+   * This is an expensive operation, and concurrent adds are permitted.
    */
   public void clear() {
-    synchronized (mLock) {
-      mObjects.clear();
-      for (Map<Object, Set<T>> mapping : mSetIndexedByFieldValue) {
-        mapping.clear();
-      }
+    for (T obj : mObjects) {
+      remove(obj);
```

Comment:
`clear()` would not respect the locking of objects, that `add()` and `remove()` respect.